src/Control/Distributed/Process/Node/RegistryAgent.hs

```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE RecursiveDo  #-}
+-----------------------------------------------------------------------------
+---- |
+---- Module      :  Control.Distributed.Process.Node.RegistryAgent
+---- Copyright   :  (c) Tweag I/O 2015
+---- License     :  BSD3 (see the file LICENSE)
+----
+---- Maintainer  :  Tim Watson <watson.timothy@gmail.com>
+---- Stability   :  experimental
+---- Portability :  non-portable (requires concurrency)
+----
+---- This module provides a registry monitoring agent, implemented as a
+---- /distributed-process Management Agent/. Every 'node' starts this agent on
+---- startup. The agent will monitor every remote process that was added to the
+---- local registry, so the node removes the process from the registry when it
+---- dies or when a network failure is detected.
+----
+-------------------------------------------------------------------------------
+
+module Control.Distributed.Process.Node.RegistryAgent
+    ( registryMonitorAgent
+    ) where
+
+import Control.Distributed.Process.Management
+import Control.Distributed.Process.Internal.Types
+import Control.Distributed.Process.Internal.Primitives
+import Data.Foldable (forM_)
+import Data.Map (Map)
+import qualified Data.Map as Map
+
+registryMonitorAgentId :: MxAgentId
+registryMonitorAgentId = MxAgentId "service.registry.monitoring"
+
+-- | Registry monitor agent
+--
+-- This agent listens for 'MxRegistered' and 'MxUnRegistered' events and tracks
+-- all remote 'ProcessId's that are stored in the registry.
+--
+-- When a remote process is registered, the agent starts monitoring it until it
+-- is unregistered or the monitor notification arrives.
+--
+-- The agent keeps the amount of labels associated with each registered remote
+-- process. This is necessary so the process is unmonitored only when it has
+-- been unregistered from all of the labels.
+--
+registryMonitorAgent :: Process ProcessId
+registryMonitorAgent = do
+  nid <- getSelfNode
+  -- For each process the map associates the 'MonitorRef' used to monitor it and
+  -- the amount of labels associated with it.
+  mxAgent registryMonitorAgentId (Map.empty :: Map ProcessId (MonitorRef, Int))
```

Comment:
Well, if events get dropped then even the current code will have trouble. Events on the management bus shouldn't get dropped.

If we must assume that messages do get reordered (because the Management API documentation claims to provide no guarantees), or dropped or duplicated, then I'm not sure that using the Management API for building this functionality is a great option after all.

As @facundominguez says, not worth agonizing over the implementation of this function, since it's basically correct as-is. Though on the assumption that the Management API does provide _some_ guarantees. @hyperthunk could you confirm what those are precisely?